torch/_refs/__init__.py

```diff
@@ -1655,48 +1655,108 @@ def item(a: TensorLikeType) -> NumberType:
+def _to_dispatch(
+    device=None,
+    dtype=None,
+    non_blocking=None,
+    copy=None,
+    memory_format=None,
+    layout=None,
+    pin_memory=None,
+):
+    if isinstance(device, TensorLike):
+        # overload to `to.other`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        # extract tensor options
+        other = device
+        device = other.device
+        dtype = other.dtype
+        layout = other.layout
+        pin_memory = other.pin_memory
+    elif isinstance(device, torch.dtype):
+        # overload to `to.dtype`
+        if isinstance(copy, torch.memory_format):
+            memory_format = copy
+        # we need to make sure that dtype is bool, otherwise, non_blocking could be passed as kwarg
+        if isinstance(non_blocking, bool) and isinstance(dtype, bool):
+            copy = non_blocking
+        if isinstance(dtype, bool):
+            non_blocking = dtype
+        dtype = device
+        layout = None
+        pin_memory = None
+        device = None
+    # overload to `to.device`
+    # do nothing because signature of 'to' matches
+    # overload to `to.dtype_layout`
+    # do nothing because the `to.dtype_layout` takes all kwargs.
+    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+
+
 def to(
     a: TensorLikeType,
-    dtype: torch.dtype,
-    non_blocking: bool = False,
-    copy: bool = False,
-    memory_format: torch.memory_format = torch.preserve_format,
+    device: Optional[torch.device] = None,
+    dtype: Optional[torch.dtype] = None,
+    non_blocking: Optional[bool] = None,
+    copy: Optional[bool] = None,
+    memory_format: Optional[torch.memory_format] = None,
+    *,
+    layout: Optional[torch.layout] = None,
+    pin_memory: Optional[bool] = None,
 ) -> TensorLikeType:
-    if _to_will_alias(a, dtype, copy, memory_format):
+    device, dtype, non_blocking, copy, memory_format, layout, pin_memory = _to_dispatch(
+        device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+    )
+
+    if _to_will_alias(a, device, dtype, copy, memory_format, layout):
         return a
     if (
-        (copy is True or dtype != a.dtype)
-        and memory_format == torch.preserve_format
-        and non_blocking is False
+        (copy or dtype != a.dtype)
+        and (memory_format is None or memory_format == torch.preserve_format)
+        and (not non_blocking)
+        and device is None
+        and layout is None
+        and pin_memory is None
     ):
         return prims.convert_element_type(a, dtype)
+    memory_format = (
+        memory_format if memory_format is not None else torch.preserve_format
```

Comment:
```suggestion
        memory_format or torch.preserve_format
```

Reply:
I prefer explicitly checking for `None` most of the time because lots of Python objects evaluate to `False`